### ISSUE TITLE
New Kokkos versions don't work with EKAT

### ISSUE THREAD

**Issue by pbosler** [E3SM-Project/EKAT#19]
Kokkos release 3.2 (and the Kokkos master branch) fail to build within EKAT.

Newer Kokkos versions enforce case-sensitive CMake variables, split libkokkos.a into 2 (libkokkoscore.a and libkokkoscontainers.a), and change some memory traits (RandomAccess) that ekat currently uses.

Alternative: Force EKAT to use the same Kokkos SHA as scream.

**Comment by bartgol** [comment]
Yes, this is an issue. So far, I put it on the back burner, since in E3SM we freeze the kokkos sha, so we are ok. But if we want ekat to be more "open" to kokkos versions, this needs to be addressed.

@jgfouca @jeff-cohere Do you think we should support different kokkos versions? If not, probably we should check the version, to make sure we're compatible.

**Comment by jeff-cohere** [comment]
From what I understand, Kokkos is being pretty actively developed, so supporting multiple versions might be a lot of work. I don't know enough to accurately estimate that work, though. Maybe we could actually get away with supporting two versions for now: the old one used by SCREAM, and the latest release.

That would buy us some time to make progress in EAGLES using a more recent version of Kokkos under our hood, but we'd still have to eventually update SCREAM's version at some point. And I think it's only fair if Pete and I lend our hands to updating SCREAM's Kokkos, since we're the ones who would arguably be most interested.

Once we get SCREAM's Kokkos updated to the latest release, though, I think we should stick to just supporting that version of Kokkos. Library versioning is a huge can of worms. :-)

**Comment by bartgol** [comment]
That reminds me: the homme-nh-xx work is using a much more recent version of kokkos. I am currently working on integrating nhxx into e3sm master, which means the upstream repo (e3sm) will soon have an updated version of kokkos...which means scream will have to upgrade as well.

Perhaps we can put this on hold, and do an upgrade of our kokkos when we merge the upstream (with the new kokkos version) soon? I will likely take care of that.

**Comment by jeff-cohere** [comment]
That sounds like a great idea, @bartgol .

**Comment by jeff-cohere** [comment]
Would a Kokkos submodule for EKAT guarantee that a valid version is always used? When SCREAM uses EKAT, would it use EKAT's version of Kokkos, or its own? If SCREAM uses its own Kokkos, separate from EKAT, what happens if the two Kokkos versions are different?

**Comment by bartgol** [comment]
SCREAM does not directly build kokkos, it delegates kokkos handling to EKAT. All SCREAM does is providing EKAT with the src path for Kokkos (since ekat doesn't know where to find it automagically, being an independent repo).

If we put kokkos _inside_ EKAT, then a) it doesn't need you to specify the kokkos path, and b) it can make sure the version of kokkos stored inside is supported.

I would do the following:

- submodule kokkos in ekat; that version is _guaranteed_ to be compiling/working
- still allow the user to pass Kokkos src dir to ekat (in case they want to use a different version); in this case, ekat tells you that "you're on your own", and we don't assume any responsibility if ekat doesn't compile/work as expected. I think this can still be useful if someone wants to quickly compare two kokkos versions, and see if there's any difference in performance (or other metric) for a specific ekat customer
- submodule yaml-cpp as well. As for kokkos, still allow the user to pass a yaml src dir, but again, as for kokkos, you're on your own.

Notice that for both kokkos and yaml we are currently using a special branch from a fork of the main repo, since in both cases we added some minor mods. For kokkos, those minor mods are now in develop, so any "reasonable" user provided kokkos src would do. For yamlcpp, that's not the case (I opened jbeder/yaml-cpp#831 but I haven't followed up, nor issued a pr for them), so a vanilla yaml-cpp version is less likely to work.

**Comment by jeff-cohere** [comment]
That sounds great, and matches the philosophy we've been following for third-party libs in our EAGLES work. It's good to know that all of SCREAM's Kokkos stuff is handled via ekat, too.

**Comment by bartgol** [comment]
The only doubts I have are on submodules duplication. I don't think there's any way to get around the kokkos duplication, since Homme needs kokkos, but doesn't use EKAT. Yaml _might_ be moved to simply be a submodule of EKAT, and removed from the main externals folder.

Note: we _could_ (and probably should, in the long run) make Homme use EKAT (and use ekat's packs, for instance, ditching the vector structure we copied from KokkosKernels). That would remove the only other customer of Kokkos that does not use EKAT. However, I wonder whether kokkos should be kept in the externals folder anyways, in case someone wants to use kokkos without using EKAT, or if we should _force_ people to use EKAT if they want ot use Kokkos...

**Comment by jeff-cohere** [comment]
Hey @bartgol , do you need any help with this submoduling stuff, or have a timeframe in mind for doing it? I'm starting to set up a clean "haero" repo for our aerosol model in which Pete and I will vet everything before it enters. If we wait on adding submodules for kokkos and yaml-cpp to the repo, I'll have to use submodules for those libraries in haero. It's not a big deal either way, but if you would like to have this done sooner in EKAT, I'd be happy to do it or to help if needed.

**Comment by bartgol** [comment]
I could do it. But I am not sure how to handle the submodule duplication. Yaml can easily be submodule _only_ in EKAT. But what about kokkos. Should kokkos be found inside `externals` as well as in `externals/ekat/extern`? Cloning the scream repo would end up cloning kokkos twice, which is kind of a waste. OTOH, moving it to be _just_ inside EKAT would force other Kokkos customers (as of now, I think that's only homme) to find kokkos inside EKAT. But modifying Homme build system requires modifying the E3SM upstream...which means EKAT would have to be in the upstream...

I don't have an answer to this dilemma, so I'm open to suggestions. The ultimate solution should be, imho, to have Kokkos _only inside_ EKAT. But is forcing other E3SM sub-projects to use EKAT (at the very least as a proxy to clone kokkos) ok? Maybe? I don't know.

**Comment by PeterCaldwell** [comment]
I'm a big fan of having a single Kokkos instance for our entire project but... I can imagine Kokkos users from outside the SCREAM ecosystem having an absolute meltdown if their impl quit working and they needed to start using EKAT, which is probably pretty orthogonal to the rest of the code they have. Do we have any such users? You all may know better than me, but I recall Luke van Roekel on the ocean side talking about Kokkos a year or two ago. Matt Norman may also have an opinion.

**Comment by jeff-cohere** [comment]
It sounds like having EKAT use a specified Kokkos installation (instead of including it as a submodule) probably provides the most flexible route. Maybe we can best pitch EKAT as a layer on top of Kokkos, and support a minimum version. If EKAT becomes A Thing with enough of the project, maybe bundling Kokkos with it would make more sense.

**Comment by bartgol** [comment]
@jeff-cohere , since EKAT is not yet in e3sm master, we should probably do what's best in our fork(s). Let's move fwd with kokkos "as part of ekat". It seems like that would be the most ideal scenario anyways.

In main e3sm, I can only think of homme as kokkos customer. And I know a guy who could accommodate this kokkos change... ;-)

If you want to go ahead and do the mod in ekat, then we can do a PR in scream to update ekat's version.

**Comment by jeff-cohere** [comment]
Okay, I'm going to add yaml-cpp and Kokkos to EKAT as submodules. Luca and I agreed to use revision 7d3eac79 from the Kokkos repo, which matches the version used for Gordon Bell prize submission, and is known to work well with our stuff. It's 6 months old, which is much newer than what SCREAM is using. It's not a release, but we'll move to an appropriate release in a future PR when we have some time to do proper performance testing.
